lib/ClangImporter/ImporterImpl.h

```diff
@@ -343,9 +376,39 @@ class LLVM_LIBRARY_VISIBILITY ClangImporter::Implementation
                  DWARFImporterDelegate *dwarfImporterDelegate);
   ~Implementation();
 
+  class DiagnosticWalker : public clang::RecursiveASTVisitor<DiagnosticWalker> {
+  public:
+    DiagnosticWalker(ClangImporter::Implementation &Impl);
+    bool TraverseDecl(clang::Decl *D);
+    bool TraverseParmVarDecl(clang::ParmVarDecl *D);
+    bool VisitDecl(clang::Decl *D);
+    bool VisitMacro(const clang::MacroInfo *MI);
+    bool VisitObjCObjectPointerType(clang::ObjCObjectPointerType *T);
+    bool VisitType(clang::Type *T);
+
+  private:
+    Implementation &Impl;
+    clang::SourceLocation TypeReferenceSourceLocation;
+  };
+
   /// Swift AST context.
   ASTContext &SwiftContext;
 
+  // Associates a vector of import diagnostics with a ClangNode
+  std::unordered_map<ImportDiagnosticTarget, std::vector<ImportDiagnostic>,
+                     ImportDiagnosticTargetHasher>
+      ImportDiagnostics;
+
+  // Tracks the set of import diagnostics already produced for deduplication
+  // purposes.
+  std::unordered_set<ImportDiagnostic, ImportDiagnosticHasher>
+      CollectedDiagnostics;
+
+  // ClangNodes for which all import diagnostics have been both collected and
+  // emitted.
+  std::unordered_set<ImportDiagnosticTarget, ImportDiagnosticTargetHasher>
+      DiagnosedValues;
```

Comment:
Can this be derived from `ImportDiagnostics`?

Reply:
No since this set tracks "has `diagnoseValue` / `diagnoseTargetDirectly` has been called for each target", not "do diagnostics exist for each target". I can't remove entries from `ImportDiagnostics` when diagnosed either, since an ancestor node might be diagnosed later. That being said, it is achievable to remove this set by having type checking not call diagnose more times than there are referencing Swift nodes. I allude to this as a possible follow up in the comment I made after submitting the lastest commit. It'll take some work since unqualified lookup is so far removed from the ClangImporter and can lookup basically anything (including regular Swift decls).